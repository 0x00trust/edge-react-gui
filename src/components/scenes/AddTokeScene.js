// @flow

import _ from 'lodash'
import * as React from 'react'
import { ActivityIndicator, Alert, ScrollView, StyleSheet, View } from 'react-native'

import { MAX_TOKEN_CODE_CHARACTERS, getSpecialCurrencyInfo } from '../../constants/indexConstants.js'
import s from '../../locales/strings.js'
import { PrimaryButton } from '../../modules/UI/components/Buttons/PrimaryButton.ui.js'
import Text from '../../modules/UI/components/FormattedText/FormattedText.ui.js'
import { THEME } from '../../theme/variables/airbitz.js'
import type { CustomTokenInfo, GuiWallet } from '../../types/types.js'
import { scale } from '../../util/scaling.js'
import { decimalPlacesToDenomination } from '../../util/utils.js'
import { FormField } from '../common/FormField.js'
import { SceneWrapper } from '../common/SceneWrapper.js'

export type AddTokenOwnProps = {
  walletId: string,
  addTokenPending: Function,
  addNewToken: Function,
  currentCustomTokens: CustomTokenInfo[],
  wallet: GuiWallet,
  onAddToken: Function,
  // adding properties in case coming from Scan scene (scan QR code to add token)
  currencyName: string,
  currencyCode: string,
  contractAddress: string,
  decimalPlaces: string
}

export type AddTokenDispatchProps = {
  addNewToken: (walletId: string, currencyName: string, currencyCode: string, contractAddress: string, denomination: string, type: string) => void
}

export type AddTokenStateProps = {
  addTokenPending: boolean,
  wallet: GuiWallet
}

type State = {
  currencyName: string,
  currencyCode: string,
  contractAddress: string,
  decimalPlaces: string,
  multiplier: string,
  enabled?: boolean
}

export type AddTokenProps = AddTokenOwnProps & AddTokenStateProps & AddTokenDispatchProps

export class AddToken extends React.Component<AddTokenProps, State> {
  constructor(props: AddTokenProps) {
    super(props)
    const { currencyName, currencyCode, contractAddress, decimalPlaces } = props
    this.state = {
      currencyName: currencyName || '',
      currencyCode: currencyCode || '',
      contractAddress: contractAddress || '',
      decimalPlaces: decimalPlaces || '',
      multiplier: ''
    }
  }

  render() {
    const { addTokenPending, wallet } = this.props
    const { currencyCode: parentCurrencyCode } = wallet
    const { isCustomTokensSupported: tokenParams } = getSpecialCurrencyInfo(parentCurrencyCode)

    return (
      <SceneWrapper background="body">
        <ScrollView style={styles.container}>
          <View style={styles.instructionalArea}>
            <Text style={styles.instructionalText}>{s.strings.addtoken_top_instructions}</Text>
          </View>
          <View style={styles.nameArea}>
            <FormField
              value={this.state.currencyName}
              onChangeText={this.onChangeName}
              autoCapitalize="words"
              autoFocus
              label={s.strings.addtoken_name_input_text}
              returnKeyType="done"
              autoCorrect={false}
            />
          </View>
          <View style={styles.currencyCodeArea}>
            <FormField
              value={this.state.currencyCode}
              onChangeText={this.onChangeCurrencyCode}
              autoCapitalize="characters"
              label={s.strings.addtoken_currency_code_input_text}
              returnKeyType="done"
              autoCorrect={false}
              maxLength={MAX_TOKEN_CODE_CHARACTERS}
            />
          </View>
          {tokenParams.contractAddress && (
            <View style={styles.contractAddressArea}>
              <FormField
                value={this.state.contractAddress}
                onChangeText={this.onChangeContractAddress}
                label={s.strings.addtoken_contract_address_input_text}
                returnKeyType="done"
                autoCorrect={false}
              />
            </View>
          )}
          {tokenParams.decimalPlaces && (
            <View style={styles.decimalPlacesArea}>
              <FormField
                value={this.state.decimalPlaces}
                onChangeText={this.onChangeDecimalPlaces}
                label={s.strings.addtoken_denomination_input_text}
                returnKeyType="done"
                autoCorrect={false}
                keyboardType="numeric"
              />
            </View>
          )}
          <View style={styles.buttonsArea}>
            <PrimaryButton style={styles.saveButton} onPress={this._onSave}>
              {addTokenPending ? <ActivityIndicator /> : <PrimaryButton.Text>{s.strings.string_save}</PrimaryButton.Text>}
            </PrimaryButton>
          </View>
          <View style={styles.bottomPaddingForKeyboard} />
        </ScrollView>
      </SceneWrapper>
    )
  }

  onChangeName = (input: string) => {
    this.setState({
      currencyName: input
    })
  }

  onChangeCurrencyCode = (input: string) => {
    this.setState({
      currencyCode: input
    })
  }

  onChangeDecimalPlaces = (input: string) => {
    this.setState({
      decimalPlaces: input
    })
  }

  onChangeContractAddress = (input: string) => {
    this.setState({
      contractAddress: input.trim()
    })
  }

  _onSave = () => {
    const currencyCode = this.state.currencyCode.toUpperCase()
    const { wallet } = this.props
    const { currencyCode: parentCurrencyCode } = wallet
    const { isCustomTokensSupported: tokenParams } = getSpecialCurrencyInfo(parentCurrencyCode)
    this.setState(
      {
        currencyCode
      },
      () => {
        const { currencyName, decimalPlaces, contractAddress } = this.state
        const { currentCustomTokens, wallet, walletId, addNewToken, onAddToken } = this.props
        const currentCustomTokenIndex = _.findIndex(currentCustomTokens, item => item.currencyCode === currencyCode)
        const metaTokensIndex = _.findIndex(wallet.metaTokens, item => item.currencyCode === currencyCode)
        // if token is hard-coded into wallets of this type
        if (metaTokensIndex >= 0) Alert.alert(s.strings.manage_tokens_duplicate_currency_code)
        // if that token already exists and is visible (ie not deleted)
        if (currentCustomTokenIndex >= 0 && currentCustomTokens[currentCustomTokenIndex].isVisible !== false) {
          Alert.alert(s.strings.manage_tokens_duplicate_currency_code)
        } else {
          // only require valid criteria
          if (
            ((tokenParams.currencyName && currencyName) || !tokenParams.currencyName) &&
            ((tokenParams.currencyCode && currencyCode) || !tokenParams.currencyCode) &&
            ((tokenParams.decimalPlaces && decimalPlaces) || !tokenParams.decimalPlaces) &&
            ((tokenParams.contractAddress && contractAddress) || !tokenParams.contractAddress)
          ) {
            const denomination = decimalPlacesToDenomination(decimalPlaces)
            console.log('addTokenScene and contractAddress is: ', contractAddress)
            addNewToken(walletId, currencyName, currencyCode, contractAddress, denomination, wallet.type)
            onAddToken(currencyCode)
          } else {
            Alert.alert(s.strings.addtoken_invalid_information)
          }
        }
      }
    )
  }
}

const rawStyles = {
  container: {
    flex: 1,
    paddingHorizontal: scale(20)
  },

  instructionalArea: {
    paddingVertical: scale(16),
    paddingHorizontal: scale(20)
  },
  instructionalText: {
    fontSize: scale(16),
    textAlign: 'center'
  },

  nameArea: {
    height: scale(70)
  },
  currencyCodeArea: {
    height: scale(70)
  },
  contractAddressArea: {
    height: scale(70)
  },
  decimalPlacesArea: {
    height: scale(70)
  },
  buttonsArea: {
    marginTop: scale(16),
    height: scale(52),
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    alignSelf: 'flex-end',
    paddingVertical: scale(4)
  },
  saveButton: {
    flex: 1,
    marginLeft: scale(2),
    backgroundColor: THEME.COLORS.SECONDARY,
    borderRadius: 3
  },
  bottomPaddingForKeyboard: {
    height: scale(300)
  }
}
const styles: typeof rawStyles = StyleSheet.create(rawStyles)
