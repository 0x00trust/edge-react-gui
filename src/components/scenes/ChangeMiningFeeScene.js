// @flow

import { type EdgeCurrencyWallet } from 'edge-core-js/types'
import React, { Component, type Node } from 'react'
import { ScrollView, Text, TouchableWithoutFeedback, View } from 'react-native'
import EntypoIcon from 'react-native-vector-icons/Entypo'
import { connect } from 'react-redux'

import { sendConfirmationUpdateTx } from '../../actions/SendConfirmationActions.js'
import { FEE_STRINGS } from '../../constants/WalletAndCurrencyConstants.js'
import s from '../../locales/strings.js'
import { getGuiMakeSpendInfo } from '../../modules/UI/scenes/SendConfirmation/selectors.js'
import { type FeeOption } from '../../reducers/scenes/SendConfirmationReducer.js'
import { dayText, nightText } from '../../styles/common/textStyles.js'
import { styles } from '../../styles/scenes/ChangeMiningFeeStyle.js'
import { THEME } from '../../theme/variables/airbitz.js'
import { type Dispatch, type State as ReduxState } from '../../types/reduxTypes.js'
import { FormField } from '../common/FormField.js'
import { SceneWrapper } from '../common/SceneWrapper.js'

type OwnProps = {
  wallet: EdgeCurrencyWallet
}

type StateProps = {
  networkFeeOption?: FeeOption,
  customNetworkFee?: Object
}

type DispatchProps = {
  onSubmit(networkFeeOption: string, customNetworkFee: Object): mixed
}

type Props = OwnProps & StateProps & DispatchProps

type State = {
  networkFeeOption: FeeOption,
  customNetworkFee: Object
}

export class ChangeMiningFee extends Component<Props, State> {
  hasCustomFees: boolean

  constructor (props: Props) {
    super(props)
    const { networkFeeOption = 'standard', customNetworkFee, wallet } = props

    // Figure out how custom fees are supposed to look (if any):
    const defaultCustomFee = {}
    if (wallet.currencyInfo.defaultSettings != null) {
      const { customFeeSettings } = wallet.currencyInfo.defaultSettings
      if (customFeeSettings != null) {
        this.hasCustomFees = true
        for (const row of customFeeSettings) {
          defaultCustomFee[row] = '0'
        }
      }
    }

    if (customNetworkFee == null || Object.keys(customNetworkFee).length !== Object.keys(defaultCustomFee).length) {
      // Use the default custom fees if we get bogus ones:
      this.state = { networkFeeOption, customNetworkFee: defaultCustomFee }
    } else {
      // Otherwise, use the custom fees from before:
      this.state = { networkFeeOption, customNetworkFee }
    }
  }

  componentWillUnmount () {
    const { networkFeeOption, customNetworkFee } = this.state
    this.props.onSubmit(networkFeeOption, customNetworkFee)
  }

  render () {
    return (
      <SceneWrapper background="body" hasTabs={false} avoidKeyboard>
        <ScrollView style={styles.content}>
          {this.renderRadioRow('high', s.strings.mining_fee_high_label_choice)}
          {this.renderRadioRow('standard', s.strings.mining_fee_standard_label_choice)}
          {this.renderRadioRow('low', s.strings.mining_fee_low_label_choice)}
          {this.hasCustomFees != null ? this.renderRadioRow('custom', s.strings.mining_fee_custom_label_choice) : null}
          {this.renderCustomFee()}
          {this.renderFeeWarning()}
        </ScrollView>
      </SceneWrapper>
    )
  }

  renderRadioRow (value: FeeOption, label: string) {
    const { networkFeeOption } = this.state

    return (
      <TouchableWithoutFeedback onPress={() => this.setState({ networkFeeOption: value })}>
        <View style={styles.radioRow}>
          <View style={[styles.radio, networkFeeOption === value ? styles.selected : null]} />
          <Text style={dayText('row-left')}>{label}</Text>
        </View>
      </TouchableWithoutFeedback>
    )
  }

  renderCustomFee (): Node {
    const { networkFeeOption, customNetworkFee } = this.state
    if (networkFeeOption !== 'custom') return null

    return (
      <View style={styles.customArea}>
        {Object.keys(customNetworkFee).map(key => (
          <FormField
            key={key}
            keyboardType="numeric"
            onChangeText={text =>
              this.setState({
                customNetworkFee: { ...customNetworkFee, [key]: text }
              })
            }
            value={customNetworkFee[key]}
            label={FEE_STRINGS[key] || key}
          />
        ))}
      </View>
    )
  }

  renderFeeWarning () {
    const { networkFeeOption } = this.state
    if (networkFeeOption !== 'custom' && networkFeeOption !== 'low') return null

    return (
      <View style={styles.warningBox}>
        <EntypoIcon name="warning" color={THEME.COLORS.WHITE} size={THEME.rem(1.4)} />
        <Text style={nightText('small')}>{s.strings.warning_low_or_custom_fee}</Text>
      </View>
    )
  }
}

export const ChangeMiningFeeScene = connect(
  (state: ReduxState): StateProps => ({
    networkFeeOption: getGuiMakeSpendInfo(state).networkFeeOption,
    customNetworkFee: getGuiMakeSpendInfo(state).customNetworkFee
  }),
  (dispatch: Dispatch): DispatchProps => ({
    onSubmit (networkFeeOption: string, customNetworkFee: Object) {
      dispatch(sendConfirmationUpdateTx({ networkFeeOption, customNetworkFee }))
    }
  })
)(ChangeMiningFee)
