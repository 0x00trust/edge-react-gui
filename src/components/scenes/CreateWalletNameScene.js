// @flow

import * as React from 'react'
import { Alert } from 'react-native'
import { sprintf } from 'sprintf-js'

import { CREATE_WALLET_REVIEW } from '../../constants/SceneKeys.js'
import s from '../../locales/strings.js'
import { type RouteProp, Actions } from '../../types/routerTypes.js'
import { SceneWrapper } from '../common/SceneWrapper'
import { MainButton } from '../themed/MainButton.js'
import { type OutlinedTextInputRef, OutlinedTextInput } from '../themed/OutlinedTextInput.js'
import { SceneHeader } from '../themed/SceneHeader'

export type CreateWalletNameOwnProps = {
  route: RouteProp<'createWalletName'>
}
type Props = CreateWalletNameOwnProps
type State = {
  walletName: string
}

export class CreateWalletName extends React.Component<Props, State> {
  textInput: { current: OutlinedTextInputRef | null } = React.createRef()

  constructor(props: Props) {
    super(props)
    const { route } = props
    const { currencyCode, currencyName } = route.params.selectedWalletType

    const name = currencyCode.toLowerCase() === 'xrp' ? 'XRP' : currencyName
    const walletName = sprintf(s.strings.my_crypto_wallet_name, name)
    this.state = { walletName }
  }

  isValidWalletName = () => {
    const { walletName } = this.state
    const isValid = walletName.length > 0

    return isValid
  }

  clearText = () => {
    this.setState({ walletName: '' })
    if (this.textInput.current) {
      this.textInput.current.blur()
    }
  }

  onNext = () => {
    const { route } = this.props
    const { cleanedPrivateKey, selectedFiat, selectedWalletType } = route.params
    if (this.isValidWalletName()) {
      Actions.push(CREATE_WALLET_REVIEW, {
        walletName: this.state.walletName,
        selectedFiat: selectedFiat,
        selectedWalletType: selectedWalletType,
        cleanedPrivateKey
      })
    } else {
      Alert.alert(s.strings.create_wallet_invalid_name, s.strings.create_wallet_enter_valid_name)
    }
  }

  handleChangeWalletName = (walletName: string) => {
    this.setState({ walletName })
  }

  render() {
    return (
      <SceneWrapper avoidKeyboard background="theme">
        <SceneHeader withTopMargin title={s.strings.title_create_wallet} />

        <OutlinedTextInput
          onChangeText={this.handleChangeWalletName}
          value={this.state.walletName}
          onSubmitEditing={this.onNext}
          autoFocus
          autoCorrect={false}
          returnKeyType="next"
          label={s.strings.fragment_wallets_addwallet_name_hint}
          onClear={this.clearText}
          clearIcon
          marginRem={[0, 1.75]}
          ref={this.textInput}
          blurOnSubmit
        />
        <MainButton alignSelf="center" label={s.strings.string_next_capitalized} marginRem={[3, 1]} type="secondary" onPress={this.onNext} />
      </SceneWrapper>
    )
  }
}
