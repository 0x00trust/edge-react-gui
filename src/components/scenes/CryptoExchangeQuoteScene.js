// @flow

import type { EdgeAccount } from 'edge-core-js'
import React, { Component } from 'react'
import { ActivityIndicator, Image, View } from 'react-native'
import { sprintf } from 'sprintf-js'

import { swapPluginLogos } from '../../assets/images/exchange'
import s from '../../locales/strings.js'
import { ExchangeQuoteComponent } from '../../modules/UI/components/ExchangeQuote/ExchangeQuoteComponent.js'
import FormattedText from '../../modules/UI/components/FormattedText/index'
import Slider from '../../modules/UI/components/Slider/index'
import { styles } from '../../styles/scenes/CryptoExchangeQuoteSceneStyles.js'
import { type GuiSwapInfo } from '../../types/types.js'
import { CircleTimer } from '../common/CircleTimer'
import { SceneWrapper } from '../common/SceneWrapper.js'
import { SwapVerifyChangellyModal } from '../modals/SwapVerifyChangellyModal.js'
import { Airship } from '../services/AirshipInstance.js'

export type OwnProps = {
  swapInfo: GuiSwapInfo
}
export type StateProps = {
  account: EdgeAccount,
  fromCurrencyIcon: string,
  fromDenomination: string,
  fromWalletCurrencyName: string,
  pending: boolean,
  toCurrencyIcon: string,
  toDenomination: string,
  toWalletCurrencyName: string
}
export type DispatchProps = {
  shift(swapInfo: GuiSwapInfo): mixed,
  timeExpired(swapInfo: GuiSwapInfo): void
}

type Props = OwnProps & StateProps & DispatchProps
type State = {}

class CryptoExchangeQuoteScreenComponent extends Component<Props, State> {
  componentDidMount = () => {
    const { swapInfo } = this.props
    const { pluginName } = swapInfo.quote
    if (pluginName === 'changelly') {
      this.checkChangellyKYC()
    }
    global.firebase && global.firebase.analytics().logEvent(`Exchange_Shift_Quote`)
  }

  renderSlider = () => {
    const { pending, shift, swapInfo } = this.props
    if (pending) {
      return <ActivityIndicator style={{ flex: 1, alignSelf: 'center' }} size={'small'} />
    }
    return <Slider onSlidingComplete={() => shift(swapInfo)} sliderDisabled={pending} parentStyle={styles.slideContainer} />
  }
  renderTimer = () => {
    const { swapInfo, timeExpired } = this.props
    const { expirationDate } = swapInfo.quote

    if (!expirationDate) return null
    return <CircleTimer style={styles.timerContainer} timeExpired={() => timeExpired(swapInfo)} expiration={expirationDate} />
  }

  async checkChangellyKYC () {
    const { account, swapInfo, timeExpired } = this.props
    const swapConfig = account.swapConfig.changelly
    if (swapConfig.userSettings && swapConfig.userSettings.agreedToTerms) return

    const result = await Airship.show(bridge => <SwapVerifyChangellyModal bridge={bridge} />)

    if (result) {
      await account.swapConfig.changelly.changeUserSettings({ agreedToTerms: true })
    } else {
      // If the user rejects the terms, disable Changelly and expire the quote:
      await account.swapConfig.changelly.changeUserSettings({ agreedToTerms: false })
      await account.swapConfig.changelly.changeEnabled(false)
      timeExpired(swapInfo)
    }
  }

  render () {
    const { fromCurrencyIcon, fromDenomination, fromWalletCurrencyName, swapInfo, toCurrencyIcon, toDenomination, toWalletCurrencyName } = this.props
    const { fee, fromDisplayAmount, fromFiat, toDisplayAmount, toFiat } = swapInfo
    const { isEstimate, pluginName } = swapInfo.quote
    const { fromWallet, toWallet } = swapInfo.request

    return (
      <SceneWrapper>
        <View style={styles.topRow}>
          <Image source={swapPluginLogos[pluginName]} resizeMode={'contain'} style={styles.logoImage} />
        </View>
        <View style={styles.centerRow}>
          <ExchangeQuoteComponent
            cryptoAmount={fromDisplayAmount}
            currency={fromWalletCurrencyName}
            currencyCode={fromDenomination}
            fiatCurrencyAmount={fromFiat}
            fiatCurrencyCode={fromWallet.fiatCurrencyCode.replace('iso:', '')}
            headline={sprintf(s.strings.exchange_will_be_sent, fromDisplayAmount, fromDenomination)}
            isTop
            miningFee={fee}
            walletIcon={fromCurrencyIcon}
            walletName={fromWallet.name || ''}
          />
          <ExchangeQuoteComponent
            cryptoAmount={toDisplayAmount}
            currency={toWalletCurrencyName}
            currencyCode={toDenomination}
            fiatCurrencyAmount={toFiat}
            fiatCurrencyCode={toWallet.fiatCurrencyCode.replace('iso:', '')}
            headline={sprintf(s.strings.exchange_will_be_received, toDisplayAmount, toDenomination)}
            isEstimate={isEstimate}
            walletIcon={toCurrencyIcon}
            walletName={toWallet.name || ''}
          />
        </View>
        <View style={styles.confirmTextRow}>
          <FormattedText style={styles.confirmText}>{s.strings.confirm_to_complete_exchange}</FormattedText>
        </View>
        <View style={styles.bottomRow}>
          {this.renderSlider()}
          {this.renderTimer()}
        </View>
      </SceneWrapper>
    )
  }
}

export { CryptoExchangeQuoteScreenComponent }
