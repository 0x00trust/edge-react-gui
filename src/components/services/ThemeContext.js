// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig.js'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
