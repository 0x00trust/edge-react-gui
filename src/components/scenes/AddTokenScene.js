// @flow

import type { EdgeCurrencyWallet } from 'edge-core-js'
import * as React from 'react'
import { KeyboardAvoidingView, Platform, TextInput, View } from 'react-native'

import { addNewToken } from '../../actions/AddTokenActions'
import s from '../../locales/strings.js'
import { useEffect, useRef, useState } from '../../types/reactHooks'
import { connect } from '../../types/reactRedux'
import type { CustomTokenInfo, GuiWallet } from '../../types/types'
import { decimalPlacesToDenomination } from '../../util/utils'
import { SceneWrapper } from '../common/SceneWrapper'
import { MismatchTokenParamsModal } from '../modals/MismatchTokenParamsModal'
import { Airship, showError } from '../services/AirshipInstance'
import { type Theme, cacheStyles, useTheme } from '../services/ThemeContext'
import { EdgeTextFieldOutlined } from '../themed/EdgeOutlinedField'
import { SceneHeader } from '../themed/SceneHeader'

export type OwnProps = {
  walletId: string,
  addTokenPending: boolean,
  currencyWallet: EdgeCurrencyWallet,
  currentCustomTokens?: CustomTokenInfo[],
  currencyInfos: any[], // TODO: replace it with real data when it will be available. Also need better variable name
  wallet: GuiWallet,
  onAddToken: (currencyCode: string) => void
}

type DispatchProps = {
  addNewToken: (walletId: string, currencyName: string, currencyCode: string, contractAddress: string, denomination: string, type: string) => void
}

type StateProps = {
  addTokenPending: boolean,
  wallet: GuiWallet,
  addTokenPending: boolean
}

type ReturnKeyType = 'next' | 'done'

type Props = OwnProps & StateProps & DispatchProps

const useReturnKeyType = (fieldsData: string[], updateCallback: (Array<'done' | 'next'>) => void) => {
  useEffect(() => {
    const returnKeyTypes: Array<'done' | 'next'> = []
    let dataEmptyIndex = -1
    let dataEmptyCount = 0

    for (let index = 0; index < fieldsData.length; index++) {
      if (fieldsData[index] === '') {
        dataEmptyCount += 1

        if (dataEmptyCount === 1) dataEmptyIndex = index
      }
    }

    for (let index = 0; index < fieldsData.length; index++) {
      returnKeyTypes[index] = dataEmptyCount === 0 || (index === dataEmptyIndex && dataEmptyCount === 1) ? 'done' : 'next'
    }

    updateCallback(returnKeyTypes)
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [...fieldsData, updateCallback])
}

export const AddToken = ({ addTokenPending, currentCustomTokens = [], currencyInfos, wallet, walletId, currencyWallet, addNewToken, onAddToken }: Props) => {
  const styles = getStyles(useTheme())
  const [currencyCode, setCurrencyCode] = useState<string>('')
  const [currencyName, setCurrencyName] = useState<string>('')
  const [contractAddress, setContractAddress] = useState<string>('')
  const [decimalPlaces, setDecimalPlaces] = useState<string>('')

  const [currencyCodeReturnKeyType, setCurrencyCodeReturnKeyType] = useState<ReturnKeyType>('next')
  const [currencyNameReturnKeyType, setCurrencyNameReturnKeyType] = useState<ReturnKeyType>('next')
  const [contractAddressReturnKeyType, setContractAddressReturnKeyType] = useState<ReturnKeyType>('next')
  const [decimalPlacesReturnKeyType, setDecimalPlacesReturnKeyType] = useState<ReturnKeyType>('next')

  const currencyCodeInputRef = useRef<TextInput>()
  const currencyNameInputRef = useRef<TextInput>()
  const contractAddressInputRef = useRef<TextInput>()
  const decimalPlacesInputRef = useRef<TextInput>()

  const [isFirstInputFocused, setIsFirstInputFocused] = useState<boolean>(true)

  const submitValidation = async () => {
    const currentCustomTokenIndex = currentCustomTokens.findIndex(item => item.currencyCode === currencyCode)
    const metaTokensIndex = wallet.metaTokens.findIndex(item => item.currencyCode === currencyCode)

    // if token is hard-coded into wallets of this type
    if (metaTokensIndex >= 0) throw new Error(s.strings.manage_tokens_duplicate_currency_code)

    // if that token already exists and is visible (ie not deleted)
    if (currentCustomTokenIndex >= 0 && currentCustomTokens[currentCustomTokenIndex].isVisible) {
      throw new Error(s.strings.manage_tokens_duplicate_currency_code)
    } else if (!currencyName || !currencyCode || !decimalPlaces || !contractAddress) {
      throw new Error(s.strings.addtoken_invalid_information)
    }
  }

  const handleSubmit = async () => {
    try {
      await submitValidation()
    } catch (error) {
      showError(error)
      return
    }

    const isExists = currencyInfos.some(
      (item: any) =>
        item.currencyCode === currencyCode ||
        item.currencyName === currencyName ||
        item.decimalPlaces === decimalPlaces ||
        item.contractAddress === contractAddress
    )

    if (!isExists) {
      const isConfirm = await Airship.show(bridge => <MismatchTokenParamsModal bridge={bridge} />)

      if (!isConfirm) return
    }

    const denomination = decimalPlacesToDenomination(decimalPlaces)

    addNewToken(walletId, currencyName, currencyCode.toUpperCase(), contractAddress, denomination, wallet.type)
    onAddToken(currencyCode.toUpperCase())
  }

  const autocompleteForm = () => {
    const currencyInfo = currencyInfos.find(
      (item: any) =>
        item.currencyCode === currencyCode ||
        item.currencyName === currencyName ||
        item.decimalPlaces === decimalPlaces ||
        item.contractAddress === contractAddress
    )

    if (currencyInfo) {
      setCurrencyCode(currencyCode || currencyInfo.currencyCode)
      setCurrencyName(currencyName || currencyInfo.currencyName)
      setDecimalPlaces(decimalPlaces || currencyInfo.decimalPlaces)
      setContractAddress(contractAddress || currencyInfo.contractAddress)
    }
  }

  const handleSubmitEditing = async () => {
    const data = [currencyCode, currencyName, contractAddress, decimalPlaces]
    const inputRefs = [currencyCodeInputRef, currencyNameInputRef, contractAddressInputRef, decimalPlacesInputRef]
    const dataEmptyIndex = data.findIndex(item => item === '')

    if (dataEmptyIndex === -1) {
      handleSubmit()
    } else {
      autocompleteForm()

      const nextFocusedField = inputRefs[dataEmptyIndex].current

      if (nextFocusedField !== null) nextFocusedField.focus()
    }
  }

  useReturnKeyType([currencyCode, currencyName, contractAddress, decimalPlaces], (returnKeyTypes: ReturnKeyType[]) => {
    setCurrencyCodeReturnKeyType(returnKeyTypes[0])
    setCurrencyNameReturnKeyType(returnKeyTypes[1])
    setContractAddressReturnKeyType(returnKeyTypes[2])
    setDecimalPlacesReturnKeyType(returnKeyTypes[3])
  })

  return (
    <SceneWrapper avoidKeyboard background="theme">
      <SceneHeader title={s.strings.title_add_token} style={styles.header} />
      <KeyboardAvoidingView style={styles.container} enabled={!isFirstInputFocused} behavior={Platform.select({ ios: 'padding', android: 'height' })}>
        <View style={[styles.container, styles.content, styles.justifyContent]}>
          <EdgeTextFieldOutlined
            ref={currencyCodeInputRef}
            autoFocus
            onFocus={() => setIsFirstInputFocused(true)}
            onBlur={() => setIsFirstInputFocused(false)}
            onChangeText={setCurrencyCode}
            value={currencyCode}
            autoCapitalize="characters"
            returnKeyType={currencyCodeReturnKeyType}
            label={s.strings.addtoken_currency_code_input_text}
            showSearchIcon={false}
            autoCorrect={false}
            marginRem={[0.5, 0.6, 1]}
            onSubmitEditing={handleSubmitEditing}
          />
          <EdgeTextFieldOutlined
            ref={currencyNameInputRef}
            onChangeText={setCurrencyName}
            value={currencyName}
            autoCapitalize="words"
            returnKeyType={currencyNameReturnKeyType}
            label={s.strings.addtoken_name_input_text}
            showSearchIcon={false}
            autoCorrect={false}
            marginRem={[0.5, 0.6, 1]}
            onSubmitEditing={handleSubmitEditing}
          />
          <EdgeTextFieldOutlined
            ref={contractAddressInputRef}
            onChangeText={setContractAddress}
            value={contractAddress}
            returnKeyType={contractAddressReturnKeyType}
            label={s.strings.addtoken_contract_address_input_text}
            showSearchIcon={false}
            autoCorrect={false}
            marginRem={[0.5, 0.6, 1]}
            onSubmitEditing={handleSubmitEditing}
          />
          <EdgeTextFieldOutlined
            ref={decimalPlacesInputRef}
            onChangeText={setDecimalPlaces}
            value={decimalPlaces}
            returnKeyType={decimalPlacesReturnKeyType}
            label={s.strings.addtoken_denomination_input_text}
            keyboardType="numeric"
            showSearchIcon={false}
            autoCorrect={false}
            marginRem={[0.5, 0.6, 0.5]}
            onSubmitEditing={handleSubmitEditing}
          />
          <View style={styles.container} />
        </View>
      </KeyboardAvoidingView>
    </SceneWrapper>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  container: {
    flex: 1
  },
  content: {
    paddingHorizontal: theme.rem(1),
    overflow: 'hidden'
  },
  justifyContent: {
    justifyContent: 'flex-end'
  },
  header: {
    marginBottom: 0
  }
}))

export const AddTokenScene = connect<StateProps, DispatchProps, OwnProps>(
  (state, ownProps) => ({
    addTokenPending: state.ui.wallets.addTokenPending,
    wallet: state.ui.wallets.byId[ownProps.walletId],
    currencyWallet: state.core.account.currencyWallets[ownProps.walletId],
    currencyInfos: [] // TODO: replace it with real data when it will be available
  }),
  dispatch => ({
    addNewToken(walletId: string, currencyName: string, currencyCode: string, contractAddress: string, denomination: string, walletType: string) {
      dispatch(addNewToken(walletId, currencyName, currencyCode, contractAddress, denomination, walletType))
    }
  })
)(AddToken)
