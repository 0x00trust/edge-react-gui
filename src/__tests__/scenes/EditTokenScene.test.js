/* globals describe it expect */
/* eslint-disable flowtype/require-valid-file-annotation */

import * as React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import { EditTokenComponent } from '../../components/scenes/EditTokenScene'
import { getTheme } from '../../components/services/ThemeContext.js'

describe('EditTokenComponent', () => {
  it('should render with loading props', () => {
    const renderer = new ShallowRenderer()

    const props = {
      route: {
        params: {
          currencyCode: 'BTC',
          metaTokens: [
            {
              currencyCode: 'BTC',
              currencyName: 'Bitcoin',
              denominations: [
                {
                  multiplier: '10000000',
                  name: 'Sats'
                }
              ]
            }
          ],
          walletId: 'myWallet'
        }
      },
      customTokens: [
        {
          currencyName: 'Shiba Inu',
          currencyCode: 'SHIB',
          contractAddress: '0x73246288',
          multiplier: '10000000',
          denomination: 'ETH',
          denominations: [
            {
              multiplier: '100000000',
              name: 'ETH'
            }
          ]
        }
      ],
      editCustomTokenProcessing: true,
      deleteCustomToken: async (walletId, currencyCode) => undefined,
      editCustomToken: (walletId, currencyName, currencyCode, contractAddress, denomination, oldCurrencyCode) => undefined,
      theme: getTheme()
    }

    const actual = renderer.render(<EditTokenComponent {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
