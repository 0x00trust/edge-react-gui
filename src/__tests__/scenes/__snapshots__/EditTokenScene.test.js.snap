// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditTokenComponent should render with loading props 1`] = `
<SceneWrapper
  avoidKeyboard={true}
  background="body"
>
  [Function]
</SceneWrapper>
`;
